pkg/hubble/container/ring_test.go

```diff
@@ -91,6 +93,79 @@ func BenchmarkTimeLibRingRead(b *testing.B) {
 	})
 }
 
+func BenchmarkReadBufferNoReaders(b *testing.B) {
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{})
+}
+
+func BenchmarkReadBufferOneReader(b *testing.B) {
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{
+		readers: 1,
+	})
+}
+
+func BenchmarkReadBufferEightReaders(b *testing.B) {
+	b.Skip() // FIXME
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{
+		readers: 8,
+	})
+}
+
+type benchmarkRingBufferOptions struct {
+	capacity capacity
+	readers  int
+}
+
+// benchmarkRingBuffer benchmarks RingBuffer's writes and reads with multiple
+// readers.
+func benchmarkRingBuffer(b *testing.B, options benchmarkRingBufferOptions) {
+	b.ReportAllocs()
+
+	if options.capacity == capacity(0) {
+		options.capacity = benchmarkCapacity
+	}
+
+	r := NewRing(options.capacity)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	readerEventsReceived := make(chan int, options.readers)
+	for i := 0; i < options.readers; i++ {
+		rr := NewRingReader(r, 0)
+		go func() {
+			// Count the number of events received and report them over a
+			// channel.
+			eventsReceived := 0
+			for {
+				if rr.NextFollow(ctx) == nil {
+					readerEventsReceived <- eventsReceived
+					return
+				}
+				eventsReceived++
+			}
+		}()
+	}
+
+	event := &v1.Event{}
+
+	b.ResetTimer()
+
+	// Write the same event b.N times.
+	for i := 0; i < b.N; i++ {
+		r.Write(event)
+	}
+
+	// Cancel all readers.
+	cancel()
+
+	// Collect results from all readers.
+	for i := 0; i < options.readers; i++ {
+		// Ensure that all readers received all events.
+		eventsReceived := <-readerEventsReceived
+		assert.Equal(b, b.N, eventsReceived)
+	}
```

Comment:
Before doing these assertions, one need to be sure that all readers are done reading (ie all goroutines have returned). This could be achieved with a [`sync.WaitGroup`](https://golang.org/pkg/sync/#WaitGroup) for instance.

Reply:
The synchronization is achieved with channels: each go routine writes its result to the channel when it is done, and the channel receive operation collects the results.